File: src/scanners/SteamSource.js
const fsp = require("fs/promises");
const fs = require("fs");

const GameDir = require("./GameDir.js");
const { vdf2js } = require("../utils/configFormats.js");

const Source = require("./Source.js");
const SteamGame = require("../games/SteamGame");

const USER_DIR = process.env["HOME"];

/**
 * Checks if a string matches any of the passed regular expressions
 * @param {string} str - The string to test
 * @param {RegExp} regexes - The regexes to test
 * @returns {boolean} - True on match, else false
 */
function strMatchAny(str, regexes){
	for (const regex of regexes){
		if (str.match(regex)){
			return true;
		}
	}
	return false;
}

class SteamSource extends Source {

	static name = "Steam";
	static gameClass = SteamGame;

	preferCache = false;

	steamDir = `${USER_DIR}/.local/share/Steam`;
	relativeImageCachePath    = "appcache/librarycache";
	relativeLibraryConfigPath = "config/libraryfolders.vdf";

	constructor(preferCache = false) {
		super();
		this.preferCache = preferCache;
	}

	/**
	 * Get steam's config data about install dirs
	 * @returns {object} - Steam's config data (install dirs)
	 * @private
	 */
	async _getLibraryConfig() {
		const _path = `${this.steamDir}/${this.relativeLibraryConfigPath}`;
		let config = await fsp.readFile(_path, { encoding: "utf-8" });
		config = vdf2js(config);
		if (!config.libraryfolders) throw new Error("Invalid steam config");
		return config;
	}

	/**
	 * Get steam install dirs from its config data
	 * @param {object} config - Steam's config data
	 * @returns {GameDir[]} - The game dirs extracted from Steam's config
	 * @private
	 */
	async _getDirs(config) {
		const libraryfolders = config.libraryfolders;
		const keys = Object.keys(libraryfolders);
		const dirs = [];
		for (const key of keys) {
			const path = libraryfolders[key].path;
			const dir = new GameDir(path);
			dirs.push(dir);
		}
		return dirs;
	}

	/**
	 * Optional step, adding images to a game
	 * @param {SteamGame} game - The game to add images to
	 * @private
	 */
	_getGameImages(game) {
		const dir = `${this.steamDir}/${this.relativeImageCachePath}`;
		const images = {
			boxArtImage: `${dir}/${game.appId}_library_600x900.jpg`,
			coverImage : `${dir}/${game.appId}_header.jpg`,
			iconImage  : `${dir}/${game.appId}_icon.jpg`,
		};
		for (const [key, value] of Object.entries(images)) {
			const imageExists = fs.existsSync(value);
			if (imageExists) {
				game[key] = value;
			}
		}
	}

	/**
	 * Optional step, adding the installed state to a game
	 * @param {Steamgame} game - The game to get the installed state
	 * @param {object} manifestData - The game's manifest
	 * @see https://github.com/lutris/lutris/blob/master/docs/steam.rst
	 * @private
	 */
	_getGameIsInstalled(game, manifestData) {
		const stateFlags = manifestData?.AppState?.StateFlags;
		if (typeof stateFlags !== "undefined") {
			const installedMask = 4;
			game.isInstalled = stateFlags & installedMask;
		}
	}

	/**
	 * Get all steam installed games
	 * @param {GameDir[]} dirs - The steam game dirs to scan for game manifests
	 * @returns {SteamGame[]} - An array of found games
	 * @private
	 */
	async _getInstalledGames(dirs) {

		const IGNORED_ENTRIES_APPIDS = [
			"221410",
			"228980",
			"1070560", // Steam Linux Runtime
		];
		const IGNORED_ENTRIES_REGEXES = [
			/^Steamworks.*/,
			/^(S|s)team ?(L|l)inux ?(R|r)untime.*/,
			/^Proton.*/
		];

		const games = [];

		for (const dir of dirs) {

			// Get all games manifests of dir
			const manDir = `${dir.path}/steamapps`;
			let entries = [];
			try { entries = await fsp.readdir(manDir); } catch (err) { continue; }
			const manifests = entries.filter(string=>(
				string.startsWith("appmanifest_") &&
				string.endsWith(".acf")
			));

			// Get info from manifests
			for (const manName of manifests) {

				const INSTALLED_MASK = 4;
				const manPath = `${manDir}/${manName}`;
				const manContent = await fsp.readFile(manPath, { encoding: "utf-8" });
				const manData = vdf2js(manContent);
				const stateFlags = manData?.AppState?.StateFlags ?? 0;

				const appid = manData?.AppState?.appid;
				const name = manData?.AppState?.name;
				const isInstalled = stateFlags & INSTALLED_MASK;
				if (!name || !appid) continue;

				// Skip duplicate games, except if the installed state of the
				// existing duplicate is worse.
				const iDup = games.findIndex(g=>g.appId === appid);
				if (iDup !== -1) {
					if (!games[iDup].isInstalled && isInstalled) continue;
					games.splice(iDup, 1);
				}

				// Skip explicitly excluded manifests
				if (IGNORED_ENTRIES_APPIDS.includes(appid)) continue;
				if (strMatchAny(name, IGNORED_ENTRIES_REGEXES)) continue;

				// Build game
				const game = new this.constructor.gameClass(appid, name, isInstalled);
				this._getGameIsInstalled(game, manData);
				this._getGameImages(game);
				games.push(game);

			}
		}

		return games;
	}

	/**
	 * Get all steam games
	 * @returns {SteamGame[]} - An array of found games
	 * @todo add support for non installed games
	 */
	async scan() {
		const lConfig = await this._getLibraryConfig();
		const dirs    = await this._getDirs(lConfig);
		const games   = await this._getInstalledGames(dirs);
		return games;
	}

}

module.exports = SteamSource;

File: gali/sources/steam_source.py
import vdf
import os
import re

from gali.sources.game_dir import GameDir
from gali.sources.source import Source
from gali.games.steam_game import SteamGame
from gali.utils.locations import HOME

def fullmatch_any(value: str, patterns: tuple[re.Pattern]) -> bool:
	"""Test if a string matches any regex in a tuple"""
	for pattern in patterns:
		if pattern.fullmatch(value):
			return True
	return False

class SteamSource(Source):

	name : str = "Steam"
	game_class : type[SteamGame] = SteamGame
	steam_dir : str = f"{HOME}/.local/share/Steam"
	rel_library_config : str = "config/libraryfolders.vdf"
	rel_image_cache : str = "appcache/librarycache"

	def get_library_config(self) -> dict:
		path = f"{self.steam_dir}/{self.rel_library_config}"
		try:
			with open(path, "r", encoding="utf-8-sig") as file:
				config = vdf.load(file)
		except Exception as err:
			raise err
		else:
			return config

	def get_game_dirs(self, config: dict) -> tuple[str]:
		paths = []
		library_folders = config["libraryfolders"]
		if library_folders is None:
			raise KeyError()
		for key in library_folders:
			entry = library_folders[key]
			path = entry.get("path", None)
			if path is None:
				continue
			paths.append(path)
		return tuple(paths)

	def get_games(self, game_dirs: tuple[str]) -> tuple[SteamGame]:

		IGNORED_APPIDS = (221410, 228980, 1070560)
		IGNORED_REGEXES = (
			re.compile("^Steamworks.*"),
			re.compile("^(S|s)team ?(L|l)inux ?(R|r)untime.*"),
			re.compile("^Proton.*"),
		)

		# Get all manifest paths
		manifests = []
		for game_dir in game_dirs:

			# Test if dir is scannable
			path = f"{game_dir}/steamapps"
			if not os.access(path, os.R_OK):
				continue

			# Scan dir
			for dirent in os.scandir(path):
				if not dirent.is_file():
					continue
				if not dirent.name.startswith("appmanifest_"):
					continue
				if not dirent.name.endswith(".acf"):
					continue
				manifests.append(dirent.path)

		games = []

		# Read manifests
		for manifest in manifests:
			
			# Get data
			try:
				with open(manifest, "r", encoding="utf-8-sig") as file:
					data = vdf.load(file)
			except Exception as err:
				raise err

			# Get interesting info
			INSTALLED_MASK = 4
			app_state = data.get("AppState", None)
			if app_state is None:
				continue
			state_flags = int(app_state.get("StateFlags", 0))
			app_id = app_state.get("appid", None)
			name = app_state.get("name", None)
			is_installed = state_flags & INSTALLED_MASK
			if (app_id is None) or (name is None):
				continue

			# Skip known non-games
			if app_id in IGNORED_APPIDS:
				continue
			if fullmatch_any(name, IGNORED_REGEXES):
				continue

			# Build game
			game = self.game_class(
				name=name,
				app_id=app_id,
				is_installed=is_installed
			)

			# If an existing game has the same app id, 
			# if the new is installed and not the old, 
			# remove the old.
			for (i, old) in enumerate(games):
				if (
					(old.app_id == game.app_id) and 
					(not old.is_insalled) and 
					game.is_installed
				):
					del games[i]

			games.append(game)

		return games

	def scan(self) -> list[SteamGame]:
		config = self.get_library_config()
		game_dirs = self.get_game_dirs(config)
		games = self.get_games(game_dirs)
		return games

class SteamFlatpakSource(SteamSource):

	name : str = "Steam (Flatpak)"
	steam_dir : str = f"{HOME}/.var/app/com.valvesoftware.Steam/.local/share/Steam"